Mathlib/CategoryTheory/Category/Preorder.lean

```diff
@@ -0,0 +1,201 @@
+/-
+Copyright (c) 2017 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Stephen Morgan, Scott Morrison, Johannes Hölzl, Reid Barton
+
+! This file was ported from Lean 3 source module category_theory.category.preorder
+! leanprover-community/mathlib commit dc6c365e751e34d100e80fe6e314c3c3e0fd2988
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.CategoryTheory.Equivalence
+import Mathlib.Order.Hom.Basic
+
+/-!
+
+# Preorders as categories
+
+We install a category instance on any preorder. This is not to be confused with the category _of_
+preorders, defined in `Order.Category.Preorder`.
+
+We show that monotone functions between preorders correspond to functors of the associated
+categories.
+
+## Main definitions
+
+* `homOfLE` and `leOfHom` provide translations between inequalities in the preorder, and
+  morphisms in the associated category.
+* `Monotone.functor` is the functor associated to a monotone function.
+
+-/
+
+
+universe u v
+
+namespace Preorder
+
+open CategoryTheory
+
+-- see Note [lower instance priority]
+/--
+The category structure coming from a preorder. There is a morphism `X ⟶ Y` if and only if `X ≤ Y`.
+
+Because we don't allow morphisms to live in `Prop`,
+we have to define `X ⟶ Y` as `ulift (plift (X ≤ Y))`.
+See `CategoryTheory.homOfLE` and `CategoryTheory.leOfHom`.
+
+See <https://stacks.math.columbia.edu/tag/00D3>.
+-/
+instance (priority := 100) smallCategory (α : Type u) [Preorder α] : SmallCategory α
+    where
+  Hom U V := ULift (PLift (U ≤ V))
+  id X := ⟨⟨le_refl X⟩⟩
+  comp := @fun X Y Z f g => ⟨⟨le_trans _ _ _ f.down.down g.down.down⟩⟩
+#align preorder.small_category Preorder.smallCategory
+
+end Preorder
+
+namespace CategoryTheory
+
+open Opposite
+
+variable {X : Type u} [Preorder X]
+
+/-- Express an inequality as a morphism in the corresponding preorder category.
+-/
+def homOfLE {x y : X} (h : x ≤ y) : x ⟶ y :=
+  ULift.up (PLift.up h)
+#align category_theory.hom_of_le CategoryTheory.homOfLE
+
+alias homOfLE ← _root_.LE.le.hom
+#align has_le.le.hom LE.le.hom
+
+@[simp]
+theorem hom_of_le_refl {x : X} : (le_refl x).hom = 𝟙 x :=
+  rfl
+#align category_theory.hom_of_le_refl CategoryTheory.hom_of_le_refl
+
+@[simp]
+theorem hom_of_le_comp {x y z : X} (h : x ≤ y) (k : y ≤ z) :
+    homOfLE h ≫ homOfLE k = homOfLE (h.trans k) :=
+  rfl
+#align category_theory.hom_of_le_comp CategoryTheory.hom_of_le_comp
+
+/-- Extract the underlying inequality from a morphism in a preorder category.
+-/
+theorem leOfHom {x y : X} (h : x ⟶ y) : x ≤ y :=
+  h.down.down
+#align category_theory.le_of_hom CategoryTheory.leOfHom
+
+alias leOfHom ← _root_.Quiver.Hom.le
+#align quiver.hom.le Quiver.Hom.le
+
+-- porting note: linter gives: "Left-hand side does not simplify, when using the simp lemma on
```

Comment:
I've seen this linter claim nonsense before. Could you please try to prove
```
example {x y : X} (h : x ≤ y) : h.hom.le = h :=
  by simp
```
If it works: please restore the simp, add a `nolint simpNF` and adjust the porting note to indicate the the linter is wrong.

Reply:
Your example seems to work, but simp fails on
```
example {x y : X} (h : x ⟶ y) : h.le.hom = h :=
  by simp
````

Reply:
pushing the nolint for the first case...

Reply:
So, for one case the linter seems to be right and not for the other?!

Reply:
apparently